### PR TITLE
own: deduplicate owner matches in search 

### PR DESCRIPTION
**Pull request by leonore**
We were still getting duplicate owner matches in search because `Path` was part of the owner key, and couldn't simply remove that because of sub-repo permissions filtering.

The sub-repo permissions filtering job has been moved prior to the Select job.

## Test plan

I ran a search for `repo:^github\.com/sourcegraph/sourcegraph$ file:.github/workflows/licenses select:file.owners`, and checked the `matches` event in the stream response.

Before (ran against s2):
```
[{"type":"person","handle":"","email":"","username":"bobheadxi","displayName":"Robert Lin","avatarURL":"https://avatars.githubusercontent.com/u/23356519?v=4"},{"type":"person","handle":"","email":"","username":"bobheadxi","displayName":"Robert Lin","avatarURL":"https://avatars.githubusercontent.com/u/23356519?v=4"}]
```

After (locally the username can’t be resolved to an actual user):
```
[{"type":"unknownOwner","handle":"bobheadxi","email":""}]
```

### CODE REVIEW

**Comment by camdencheek** [comment]
Hi! I don't quite understand why path needs to be part of the owner key. Does sub-repo permissions use the key (I don't see any place it does)? The owners type can have information in it (like path) that isn't used as part of its key (which is specifically intended to identify duplicates). Also, if we only need the path for sub-repo permissions, then we should ideally be doing that filtering at a lower level.

Second question: is it okay to just drop duplicate owners? If the path information is actually useful, I imagine we'd want to aggregate it, right?

**Comment by limitedmage** [comment]
> Second question: is it okay to just drop duplicate owners? If the path information is actually useful, I imagine we'd want to aggregate it, right?

I think we will need to eventually aggregate the path, but we don't need to do that for Starship MVP. I agree that dropping owners works for now.

**Comment by leonore** [comment]
> Hi! I don't quite understand why path needs to be part of the owner key. 

It doesn't, which is why I've removed it.

> Does sub-repo permissions use the key (I don't see any place it does)? The owners type can have information in it (like path) that isn't used as part of its key (which is specifically intended to identify duplicates). 

Yep, so that's what I'm trying to do here.

>  Also, if we only need the path for sub-repo permissions, then we should ideally be doing that filtering at a lower level.

Right now we only need the path for sub-repo permissions. Can you clarify what you mean here? 

> Second question: is it okay to just drop duplicate owners? If the path information is actually useful, I imagine we'd want to aggregate it, right?

Juliana answered that as I would have 😄

File: internal/search/codeownership/select_job_test.go [hunk 1]
```diff
@@ -163,9 +163,6 @@ func TestGetCodeOwnersFromMatches(t *testing.T) {
 		})
 		autogold.Want("owner matches are as expected", want).Equal(t, matches)
 	})
-	t.Run("deduplicates owner matches", func(t *testing.T) {
```
Comment (eseliger) [inline internal/search/codeownership/select_job_test.go:166]: do we not want to implement such a test?

**Review by eseliger — APPROVED** [review]
sorry I didn't see this earlier, I somehow wasn't part of @sourcegraph/own 😆 

not a search expert but looks good to my non-expert eyes :)

**Comment by camdencheek** [comment]
Sorry, still a bit confused 🙂 It looks like we do [this exact same deduping](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@57831b3f77248baea9bb6a507fe6e97c83549576/-/blob/internal/search/codeownership/select_job.go?L47-52) in the `selectOwnersJob`. I still don't quite see why we need to reapply it.

**Comment by camdencheek** [comment]
Okay -- I _think_ I see the issue. 

The select owners job might emit two different matches: `match1 := Owner{Author: Camden, Path: /path/a}` and `match2 := Owner{Author: Camden, Path: /path/b}`. Those will be deduplicated in the select owners job to only `match1`. 

However, the subrepo permissions job might filter out `match1`, but wouldn't filter out `match2`. So we don't want to deduplicate until _after_ we filter using subrepo permissions.

File: internal/search/codeownership/dedup_job.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		errs error
+	)
+	dedup := result.NewDeduper()
+
+	filteredStream := streaming.StreamFunc(func(event streaming.SearchEvent) {
+		if err != nil {
+			mu.Lock()
+			errs = errors.Append(errs, err)
+			mu.Unlock()
+		}
```
Comment (camdencheek) [inline internal/search/codeownership/dedup_job.go:44]:
```suggestion
```

**Comment by leonore** [comment]
> However, the subrepo permissions job might filter out `match1`, but wouldn't filter out `match2`. So we don't want to deduplicate until _after_ we filter using subrepo permissions.

yep, you got it.

> Except I think we are still deduplicating in `SelectOwnersJob` after this change, so I don't think this does quite what you're looking for?

I think I just forgot to remove it there ha, good catch 🤦 

> I think maybe a better solution to this would be to simply move the subrepo permissions filter to before the select job.

so yeah that would be the perfect solution. I just wasn't sure whether there were some cases with sub-repo filtering we might miss with this order. Working through the logic with what I know I think there might be no issues, but I don't know whether there's edge cases with other `select` predicates I might be missing.

```
search returns files with paths A, B, C

sub-repo permissions filtering: user has no access to B

select:owners on files {A, C}

resolve firstOwner for A, firstOwner for C. 
// firstOwner might also own path B but we check ownership by file path so we wouldn't leak that information. 
```

cc @camdencheek

**Comment by camdencheek** [comment]
> Seems sound enough though

Agreed. The filter in the search pipeline is kinda a "last ditch" effort anyways. Permissions _should_ be applied closer to the data. Your outline there looks great 👍

File: internal/search/codeownership/select_job_test.go (hunk 1 above)
Comment (leonore) [reply internal/search/codeownership/select_job_test.go:166]: we would, but the deduping is done a level above in the job run function, which is harder to test

Comment (eseliger) [reply internal/search/codeownership/select_job_test.go:166]: 👍

**Review by camdencheek — APPROVED** [review]
Nice, looks good to me! Thanks for iterating with me on this one

**Comment by eseliger** [comment]
@leonore I've merged latest main trying to fix the build here!

File: internal/search/job/jobutil/job.go [hunk 3]
```diff
@@ -191,24 +191,24 @@ func NewBasicJob(inputs *search.Inputs, b query.Basic) (job.Job, error) {
...
+
 	{ // Apply selectors
 		if v, _ := b.ToParseTree().StringValue(query.FieldSelect); v != "" {
 			sp, _ := filter.SelectPathFromString(v) // Invariant: select already validated
 			basicJob = NewSelectJob(sp, basicJob)
 			// the select owners job is ran separately as it requires state and can return multiple owners from one match.
 			if inputs.Features.CodeOwnershipSearch && sp.Root() == filter.File && len(sp) == 2 && sp[1] == "owners" {
-				basicJob = codeownershipjob.NewSelectOwnersSearch(basicJob)
+				basicJob = codeownershipjob.NewSelectOwners(basicJob)
 			}
```
Comment (eseliger) [inline internal/search/job/jobutil/job.go:208]: is it intentional that we do _both_ SelectJob and SelectOwners? Or could we do

```go
			// the select owners job is ran separately as it requires state and can return multiple owners from one match.
			if inputs.Features.CodeOwnershipSearch && sp.Root() == filter.File && len(sp) == 2 && sp[1] == "owners" {
				basicJob = codeownershipjob.NewSelectOwners(basicJob)
			} else {
				basicJob = NewSelectJob(sp, basicJob)
			}
```

Comment (leonore) [reply internal/search/job/jobutil/job.go:208]: I think we could! I merged this to fix the dedup, but I've replied on the other PR as well :)

**Pull request merged**